### PR TITLE
TGUI: Microwave

### PR DESCRIPTION
**Pull request by Crowbar764**
# About the pull request

Gives microwaves a TGUI.

image

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

image

image

image

image

</details>

# Changelog

:cl: Casper
ui: added microwave TGUI
/:cl:

### CODE REVIEW

File: code/game/machinery/kitchen/microwave.dm [hunk 1]
```diff
@@ -154,72 +154,81 @@
...
-				display_name = "Hotsauce"
-			if (R.id == "frostoil")
-				display_name = "Coldsauce"
-			dat += {"<B>[display_name]:</B> [R.volume] unit\s<BR>"}
-
-		if (items_counts.len==0 && reagents.reagent_list.len==0)
-			dat = {"<B>The microwave is empty</B><BR>"}
+	data["operating"] = src.operating
+	data["broken"] = src.broken
+	data["dirty"] = (src.dirty == 100)
```
Comment (Zonespace27) [inline code/game/machinery/kitchen/microwave.dm:173]: `src.` not needed

File: code/game/machinery/kitchen/microwave.dm [hunk 2]
```diff
@@ -154,72 +154,81 @@
...
+
+	var/list/ingredients = new
+	var/list/items_counts = new
+	var/list/items_measures = new
+	var/list/items_measures_p = new
+
+	for (var/obj/O in contents)
```
Comment (Zonespace27) [inline code/game/machinery/kitchen/microwave.dm:180]: no one letter vars, add `as anything`

File: code/game/machinery/kitchen/microwave.dm [hunk 3]
```diff
@@ -154,72 +154,81 @@
...
-		if (items_counts.len==0 && reagents.reagent_list.len==0)
-			dat = {"<B>The microwave is empty</B><BR>"}
+	data["operating"] = src.operating
+	data["broken"] = src.broken
+	data["dirty"] = (src.dirty == 100)
+
+	var/list/ingredients = new
+	var/list/items_counts = new
+	var/list/items_measures = new
+	var/list/items_measures_p = new
```
Comment (Zonespace27) [inline code/game/machinery/kitchen/microwave.dm:178]: use `= list()`

File: code/game/machinery/kitchen/microwave.dm [hunk 4]
```diff
@@ -154,72 +154,81 @@
...
+
+	for (var/obj/O in contents)
+		var/display_name = O.name
+
+		if (istype(O,/obj/item/reagent_container/food/snacks/egg))
```
Comment (Zonespace27) [inline code/game/machinery/kitchen/microwave.dm:183]: spacing between comma and next arg

File: code/game/machinery/kitchen/microwave.dm [hunk 5]
```diff
@@ -154,72 +154,81 @@
...
+		if (istype(O,/obj/item/reagent_container/food/snacks/donkpocket))
+			display_name = "Turnovers"
+			items_measures[display_name] = "turnover"
+			items_measures_p[display_name] = "turnovers"
+		if (istype(O,/obj/item/reagent_container/food/snacks/carpmeat))
+			items_measures[display_name] = "fillet of meat"
+			items_measures_p[display_name] = "fillets of meat"
+		items_counts[display_name]++
+
+	for (var/O in items_counts)
```
Comment (Zonespace27) [inline code/game/machinery/kitchen/microwave.dm:201]: no one letter vars

File: code/game/machinery/kitchen/microwave.dm [hunk 6]
```diff
@@ -154,72 +154,81 @@
...
+		var/list/item = new
```
Comment (Zonespace27) [inline code/game/machinery/kitchen/microwave.dm:202]: = list()

File: code/game/machinery/kitchen/microwave.dm [hunk 7]
```diff
@@ -154,72 +154,81 @@
...
-<A href='?src=\ref[src];action=dispose'>Eject ingredients!<BR>\
-"}
+			item["measure"] = items_measures_p[O]
 
-	show_browser(user, dat, "Microwave Controls", "microwave")
-	return
+		ingredients += list(item)
 
+	for (var/datum/reagent/R in reagents.reagent_list)
+		var/list/reagent = new
```
Comment (Zonespace27) [inline code/game/machinery/kitchen/microwave.dm:217]: = list()

File: code/game/machinery/kitchen/microwave.dm [hunk 8]
```diff
@@ -154,72 +154,81 @@
...
 
+		reagent["count"] = R.volume
+		reagent["measure"] = R.volume > 1 ? "units" : "unit"
+
+		reagent["name"] = R.name
+		if (R.id == "hotsauce")
+			reagent["name"] = "Hotsauce"
+		if (R.id == "frostoil")
+			reagent["name"] = "Coldsauce"
```
Comment (Zonespace27) [inline code/game/machinery/kitchen/microwave.dm:226]: this seems kind of gross.

File: tgui/packages/tgui/interfaces/Microwave.js
Comment (Zonespace27) [inline tgui/packages/tgui/interfaces/Microwave.js:1]: Not explicitly required, but making this a typescript file would be preferred

File: tgui/packages/tgui/interfaces/Microwave.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+              </Button>
+            </Flex>
+          }>
+          {!!operating && (
+            <NoticeBox
+              width="100%"
+              textAlign="center"
+              p=".5rem"
+              fontSize="1.5rem">
+              Cooking..
```
Comment (Zonespace27) [inline tgui/packages/tgui/interfaces/Microwave.js:40]:
```suggestion
              Cooking...
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/game/machinery/kitchen/microwave.dm [hunk 10]
```diff
@@ -154,72 +154,81 @@
...
+/obj/structure/machinery/microwave/tgui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "Microwave", "Microwave Controls")
+		ui.open()
+
 //*******************
 //*   Microwave Menu
 //********************/
+/obj/structure/machinery/microwave/ui_data(mob/user)
```
[the remark this answers is not in the document]
Comment (Crowbar764) [reply code/game/machinery/kitchen/microwave.dm:168]: This seems to have the side effect of showing a loading screen every update. Maybe I'm doing something wrong or is this normal? Tried with both update_static_data(user) and update_static_data_for_all_viewers()

https://github.com/cmss13-devs/cmss13/assets/59719612/13c1045a-8da9-47f1-b137-69d4c1eadf7e

File: code/game/machinery/kitchen/microwave.dm (hunk 8 above)
Comment (Crowbar764) [reply code/game/machinery/kitchen/microwave.dm:226]: Most of it is just copy pasted from the old code. The alternative is to have it called 'Frost oil' instead, I can remove it if you want.

File: code/game/machinery/kitchen/microwave.dm (hunk 10 above)
[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/game/machinery/kitchen/microwave.dm]: That does look pretty bad, I agree. Feel free to ignore this request.

File: code/game/machinery/kitchen/microwave.dm (hunk 8 above)
Comment (Zonespace27) [reply code/game/machinery/kitchen/microwave.dm]: it can be left

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by Crowbar764** [comment]
@Zonespace27 I am happy with this if you are

**Review by Zonespace27 — APPROVED** [review]
lgtm

**Pull request merged**